openedx/core/djangoapps/notifications/views.py

```diff
@@ -63,14 +71,19 @@ def list(self, request, *args, **kwargs):
         Returns the list of active course enrollments for which ENABLE_NOTIFICATIONS
         Waffle flag is enabled
         """
-        enrollment_queryset = self.get_queryset().select_related('course')
-        enrollments = [
-            enrollment
-            for enrollment in enrollment_queryset
-            if ENABLE_NOTIFICATIONS.is_enabled(enrollment.course.id)
-        ]
-        serializer = self.get_serializer(enrollments, many=True)
-        return Response(serializer.data)
+        queryset = self.filter_queryset(self.get_queryset())
+        course_ids = queryset.order_by('-id').values_list('course_id', flat=True)
+
+        for course_id in course_ids:
+            if not ENABLE_NOTIFICATIONS.is_enabled(course_id):
+                queryset = queryset.exclude(course_id=course_id)
+
+        page = self.paginate_queryset(queryset)
```

Comment:
Before paginating the queryset we need to order_by('-id') and also select_related('course')